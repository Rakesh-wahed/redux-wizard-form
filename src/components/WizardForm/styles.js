import styled from 'styled-components';

export const Wrapper = styled.div`
  overflow: hidden;
  white-space: nowrap;
  width: 100%;
`;
