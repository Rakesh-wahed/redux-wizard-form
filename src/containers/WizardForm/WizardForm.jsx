import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { WizardSteps } from '../../components';

export const reduxWizardForm = formOptions => steps => {
  if (!Array.isArray(steps)) {
    return () => null;
  }

  class WizardForm extends Component {
    state = {
      formData: {},
      currentStep: 0,
      activatedSteps: [0]
    };

    onNextStep = () => {
      const nextStep = this.state.currentStep + 1;

      if (this.isValidStep(nextStep)) {
        this.setState({ currentStep: nextStep });
      }
    };

    onPreviousStep = () => {
      const nextStep = this.state.currentStep - 1;

      if (this.isValidStep(nextStep)) {
        this.setState({ currentStep: nextStep });
      }
    };

    onGoToStep = step => {
      if (this.isValidStep(step)) {
        this.setState({ currentStep: step });
      }
    };

    onSubmit = data => {
      const { formData, activatedSteps, currentStep } = this.state;
      const nextFormData = { ...formData, ...data };

      if (this.state.currentStep === this.stepSize - 1) {
        return this.props.onWizardComplete(nextFormData);
      }

      this.setState({
        formData: nextFormData,
        activatedSteps: [...activatedSteps, currentStep + 1]
      });

      return this.onNextStep();
    };

    stepSize = steps.length;

    isValidStep(step) {
      return step >= 0 && step < this.stepSize;
    }

    render() {
      const { currentStep, activatedSteps, nextAnimation, prevAnimation } = this.state;

      return (
        <WizardSteps
          formOptions={formOptions}
          steps={steps}
          currentStep={currentStep}
          activatedSteps={activatedSteps}
          nextAnimation={nextAnimation}
          prevAnimation={prevAnimation}
          onSubmit={this.onSubmit}
          onGoToStep={this.onGoToStep}
          onNextStep={this.onNextStep}
          onPreviousStep={this.onPreviousStep}
          {...this.props}
        />
      );
    }
  }

  WizardForm.propTypes = {
    onWizardComplete: PropTypes.func.isRequired
  };

  return WizardForm;
};
