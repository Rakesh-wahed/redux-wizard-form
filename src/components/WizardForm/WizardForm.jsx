import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Wrapper } from './styles';

export class WizardFormComponent extends Component {
  componentDidMount() {
    const { onWizardOptionsLoad, reduxFormOptions } = this.props;
    onWizardOptionsLoad(reduxFormOptions);
  }

  componentWillReceiveProps(nextProps) {
    const { onWizardComplete } = this.props;

    if (nextProps.isWizardComplete) {
      onWizardComplete(nextProps.data);
    }
  }

  componentWillUnmount() {
    this.props.wizardReset();
  }

  render() {
    const { children, isLoaded } = this.props;

    if (!isLoaded) {
      return null;
    }

    return <Wrapper>{children}</Wrapper>;
  }
}

WizardFormComponent.propTypes = {
  reduxFormOptions: PropTypes.shape({
    form: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    onSubmit: PropTypes.func,
    onSubmitFail: PropTypes.func,
    onSubmitSuccess: PropTypes.func
  }).isRequired,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
  data: PropTypes.shape({}),
  isWizardComplete: PropTypes.bool.isRequired,
  isLoaded: PropTypes.bool.isRequired,
  wizardReset: PropTypes.func.isRequired,
  onWizardOptionsLoad: PropTypes.func.isRequired,
  onWizardComplete: PropTypes.func.isRequired
};

WizardFormComponent.defaultProps = {
  data: {}
};
