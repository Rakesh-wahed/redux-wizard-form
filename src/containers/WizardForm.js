import { connect } from 'react-redux';
import { wizardLoad } from '../actions';
import { getIsLoaded, getCurrentStep, getStepsSize, getData } from '../reducer';
import { WizardFormComponent } from '../components/WizardForm/WizardForm.jsx';

export const WizardForm = connect(
  state => ({
    isLoaded: getIsLoaded(state),
    data: getData(state),
    currentStep: getCurrentStep(state),
    stepsSize: getStepsSize(state)
  }),
  dispatch => ({
    onWizardLoad: (stepsSize, formOptions) => dispatch(wizardLoad(stepsSize, formOptions))
  })
)(WizardFormComponent);
