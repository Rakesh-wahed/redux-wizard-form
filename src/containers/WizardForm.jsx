import { connect } from 'react-redux';
import { isWizardComplete, isLoaded, getData } from '../reducer';
import { WizardFormComponent } from '../components/WizardForm/WizardForm';
import { formOptionsSet, wizardReset } from '../actions';

export const WizardForm = connect(
  state => ({
    isWizardComplete: isWizardComplete(state),
    isLoaded: isLoaded(state),
    data: getData(state)
  }),
  dispatch => ({
    wizardReset: () => dispatch(wizardReset()),
    onWizardOptionsLoad: formOptions => dispatch(formOptionsSet(formOptions))
  })
)(WizardFormComponent);
