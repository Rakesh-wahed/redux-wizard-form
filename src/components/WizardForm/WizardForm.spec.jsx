import React from 'react';
import 'raf/polyfill';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { WizardFormComponent } from './WizardForm';

Enzyme.configure({ adapter: new Adapter() });

describe('WizardForm component', () => {
  test('snapshot', () => {
    const props = {
      reduxFormOptions: { form: 'formTest' },
      isWizardComplete: false,
      isLoaded: true,
      wizardReset: () => {},
      onWizardOptionsLoad: () => {},
      onWizardComplete: () => {}
    };
    const Component = shallow(
      <WizardFormComponent {...props}>
        <p> Test </p>
      </WizardFormComponent>
    );
    expect(Component).toMatchSnapshot();
  });
});
