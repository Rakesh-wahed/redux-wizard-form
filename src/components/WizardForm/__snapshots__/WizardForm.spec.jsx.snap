// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardForm component snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WizardFormComponent
    data={Object {}}
    isLoaded={true}
    isWizardComplete={false}
    onWizardComplete={[Function]}
    onWizardOptionsLoad={[Function]}
    reduxFormOptions={
      Object {
        "form": "formTest",
      }
    }
    wizardReset={[Function]}
  >
    <p>
       Test 
    </p>
  </WizardFormComponent>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <p>
         Test 
      </p>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": " Test ",
      },
      "ref": null,
      "rendered": " Test ",
      "type": "p",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <p>
           Test 
        </p>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " Test ",
        },
        "ref": null,
        "rendered": " Test ",
        "type": "p",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
